File: ChattersApp/src/screens/ProfileScreen.js
import { useEffect, useState } from "react";
import {
  ActivityIndicator,
  Alert,
  Pressable,
  ScrollView,
  StyleSheet,
  Text,
  TextInput,
  View,
} from "react-native";
import * as ImagePicker from "expo-image-picker";

import Screen from "../components/Screen";
import Avatar from "../components/Avatar";
import { api } from "../api/client";
import { getMe } from "../api/auth";
import { deleteAvatar, setAvatarVisibility, uploadAvatar } from "../api/avatar";
import { loadIdentity } from "../crypto/keystore";
import { generateIdentity, wrapIdentity, isSupported } from "../crypto/e2ee";
import {
  registerForPush,
  unregisterForPush,
  getStoredPushToken,
} from "../services/notifications";
import { useAuth } from "../context/AuthContext";
import { useTheme } from "../theme";

function Section({ title, children }) {
  const theme = useTheme();
  return (
    <View style={[styles.section, { backgroundColor: theme.card, borderColor: theme.border }]}>
      <Text style={[styles.sectionTitle, { color: theme.subtext }]}>{title}</Text>
      {children}
    </View>
  );
}

export default function ProfileScreen() {
  const theme = useTheme();
  const { user: username, signOut } = useAuth();

  const [newUsername, setNewUsername] = useState("");
  const [oldPassword, setOldPassword] = useState("");
  const [newPassword, setNewPassword] = useState("");
  const [error, setError] = useState("");
  const [message, setMessage] = useState("");
  const [loading, setLoading] = useState(false);

  const [hasKey, setHasKey] = useState(false);
  const [pushOn, setPushOn] = useState(false);
  const [pushBusy, setPushBusy] = useState(false);

  const [hasAvatar, setHasAvatar] = useState(false);
  const [visibility, setVisibility] = useState("public");
  const [avatarBusy, setAvatarBusy] = useState(false);
  const [avatarBust, setAvatarBust] = useState(0);

  useEffect(() => {
    loadIdentity(username).then((id) => setHasKey(Boolean(id)));
    getStoredPushToken().then((t) => setPushOn(Boolean(t)));
    getMe()
      .then((me) => {
        setHasAvatar(Boolean(me.has_avatar));
        setVisibility(me.avatar_visibility || "public");
      })
      .catch(() => {});
  }, [username]);

  function note(msg) {
    setMessage(msg);
    setError("");
  }

  async function pickAvatar() {
    const res = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ["images"],
      quality: 0.85,
    });
    if (res.canceled) return;
    const asset = res.assets[0];
    setAvatarBusy(true);
    setError("");
    try {
      await uploadAvatar({ uri: asset.uri, name: asset.fileName, mimeType: asset.mimeType });
      setHasAvatar(true);
      setAvatarBust((n) => n + 1);
      note("Profile photo updated");
    } catch (err) {
      setError(err.message);
    } finally {
      setAvatarBusy(false);
    }
  }

  async function removeAvatar() {
    setAvatarBusy(true);
    try {
      await deleteAvatar();
      setHasAvatar(false);
      setAvatarBust((n) => n + 1);
      note("Profile photo removed");
    } catch (err) {
      setError(err.message);
    } finally {
      setAvatarBusy(false);
    }
  }

  async function changeVisibility(next) {
    const prev = visibility;
    setVisibility(next);
    try {
      await setAvatarVisibility(next);
    } catch (err) {
      setVisibility(prev);
      setError(err.message);
    }
  }

  async function togglePush() {
    setPushBusy(true);
    setError("");
    try {
      if (pushOn) {
        await unregisterForPush();
        setPushOn(false);
        note("Notifications turned off");
      } else {
        const token = await registerForPush();
        if (!token) {
          setError("Notification permission was not granted.");
        } else {
          setPushOn(true);
          note("Notifications turned on");
        }
      }
    } catch (err) {
      setError(err.message);
    } finally {
      setPushBusy(false);
    }
  }

  async function changeUsername() {
    if (!newUsername.trim()) return setError("Please enter a new username");
    setLoading(true);
    setError("");
    try {
      await api.put("/api/profile/username", { new_username: newUsername.trim() });
      Alert.alert("Username updated", "Please sign in again.", [
        { text: "OK", onPress: signOut },
      ]);
    } catch (err) {
      setError(err.message);
    } finally {
      setLoading(false);
    }
  }

  async function changePassword() {
    if (!oldPassword || !newPassword) return setError("Please fill both password fields");
    setLoading(true);
    setError("");
    try {
      let keys;
      if (isSupported()) {
        try {
          const stored = await loadIdentity(username);
          if (stored) keys = await wrapIdentity(await generateIdentity(), newPassword);
        } catch {
          keys = undefined;
        }
      }
      await api.put("/api/profile/password", {
        old_password: oldPassword,
        new_password: newPassword,
        keys,
      });
      Alert.alert("Password updated", "Please sign in again.", [
        { text: "OK", onPress: signOut },
      ]);
    } catch (err) {
      setError(err.message);
    } finally {
      setLoading(false);
    }
  }

  const f = { backgroundColor: theme.bg, borderColor: theme.border, color: theme.text };

  return (
    <Screen edges={["top"]}>
      <ScrollView contentContainerStyle={styles.content}>
        <Text style={[styles.h1, { color: theme.text }]}>Profile</Text>

        <Section title="نام کاربری">
          <Text style={{ fontSize: 16, fontWeight: "500", color: theme.text }}>{username}</Text>
        </Section>

        <Section title="📷 Profile photo">
          <View style={{ flexDirection: "row", alignItems: "center", gap: 14 }}>
            <Avatar userId={username} size={64} bust={avatarBust} />
            <View style={{ flex: 1, gap: 6 }}>
              <Pressable
                style={[styles.btnSecondary, { borderColor: theme.border }]}
                onPress={pickAvatar}
                disabled={avatarBusy}
              >
                <Text style={{ color: theme.text }}>
                  {avatarBusy ? "…" : hasAvatar ? "Change photo" : "Upload photo"}
                </Text>
              </Pressable>
              {hasAvatar && (
                <Pressable
                  style={[styles.btnSecondary, { borderColor: theme.border }]}
                  onPress={removeAvatar}
                  disabled={avatarBusy}
                >
                  <Text style={{ color: theme.text }}>Remove photo</Text>
                </Pressable>
              )}
            </View>
          </View>
          <View style={{ flexDirection: "row", gap: 8, marginTop: 10 }}>
            {["public", "contacts"].map((v) => (
              <Pressable
                key={v}
                onPress={() => changeVisibility(v)}
                style={[
                  styles.chip,
                  {
                    backgroundColor: visibility === v ? theme.primary : theme.bg,
                    borderColor: theme.border,
                  },
                ]}
              >
                <Text style={{ color: visibility === v ? "#fff" : theme.text, fontSize: 13 }}>
                  {v === "public" ? "Everyone" : "My contacts only"}
                </Text>
              </Pressable>
            ))}
          </View>
        </Section>

        <Section title="🔔 Notifications">
          <Pressable
            style={[styles.btn, { backgroundColor: pushOn ? theme.bg : theme.primary, borderColor: theme.border, borderWidth: pushOn ? StyleSheet.hairlineWidth : 0 }]}
            onPress={togglePush}
            disabled={pushBusy}
          >
            {pushBusy ? (
              <ActivityIndicator color={pushOn ? theme.text : "#fff"} />
            ) : (
              <Text style={{ color: pushOn ? theme.text : "#fff", fontWeight: "600" }}>
                {pushOn ? "Turn off notifications" : "Turn on notifications"}
              </Text>
            )}
          </Pressable>
        </Section>

        <Section title="🔒 Encryption">
          <Text style={{ color: theme.text, fontSize: 14 }}>
            {hasKey
              ? "Your encryption key is unlocked on this device. Secure chats open normally."
              : "No encryption key on this device. Sign out and back in to unlock secure chats."}
          </Text>
        </Section>

        <Section title="تغییر نام کاربری">
          <TextInput
            style={[styles.field, f]}
            placeholder="نام کاربری جدید"
            placeholderTextColor={theme.subtext}
            autoCapitalize="none"
            value={newUsername}
            onChangeText={setNewUsername}
          />
          <Pressable style={[styles.btn, { backgroundColor: theme.primary }]} onPress={changeUsername} disabled={loading}>
            <Text style={styles.btnText}>بروزرسانی نام کاربری</Text>
          </Pressable>
        </Section>

        <Section title="تغییر رمز عبور">
          <TextInput
            style={[styles.field, f]}
            placeholder="رمز عبور قدیم"
            placeholderTextColor={theme.subtext}
            secureTextEntry
            value={oldPassword}
            onChangeText={setOldPassword}
          />
          <TextInput
            style={[styles.field, f]}
            placeholder="رمز عبور جدید (حداقل ۸ کاراکتر)"
            placeholderTextColor={theme.subtext}
            secureTextEntry
            value={newPassword}
            onChangeText={setNewPassword}
          />
          <Text style={{ color: theme.subtext, fontSize: 12 }}>
            Changing your password issues a new encryption key. Existing secure chats stay
            readable only where your old key is still stored.
          </Text>
          <Pressable style={[styles.btn, { backgroundColor: theme.primary }]} onPress={changePassword} disabled={loading}>
            <Text style={styles.btnText}>بروزرسانی رمز عبور</Text>
          </Pressable>
        </Section>

        {error ? <Text style={{ color: theme.danger }}>{error}</Text> : null}
        {message ? <Text style={{ color: theme.subtext }}>{message}</Text> : null}

        <Pressable style={[styles.btn, { backgroundColor: theme.danger }]} onPress={signOut}>
          <Text style={styles.btnText}>خروج</Text>
        </Pressable>
      </ScrollView>
    </Screen>
  );
}

const styles = StyleSheet.create({
  content: { padding: 16, gap: 14 },
  h1: { fontSize: 22, fontWeight: "700" },
  section: {
    borderRadius: 14,
    borderWidth: StyleSheet.hairlineWidth,
    padding: 14,
    gap: 10,
  },
  sectionTitle: { fontSize: 13 },
  field: {
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 10,
    paddingHorizontal: 12,
    paddingVertical: 11,
    fontSize: 15,
  },
  btn: { borderRadius: 10, paddingVertical: 12, alignItems: "center" },
  btnText: { color: "#fff", fontWeight: "600" },
  btnSecondary: {
    borderRadius: 10,
    borderWidth: StyleSheet.hairlineWidth,
    paddingVertical: 10,
    alignItems: "center",
  },
  chip: {
    paddingHorizontal: 12,
    paddingVertical: 8,
    borderRadius: 20,
    borderWidth: StyleSheet.hairlineWidth,
  },
});

File: flutter/chatters_shared/lib/src/ui/screens/profile_screen.dart
import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';

import '../../api/endpoints.dart';
import '../../config.dart';
import '../../crypto/e2ee.dart';
import '../../crypto/keystore.dart';
import '../../services/auth.dart';
import '../theme.dart';
import '../widgets/avatar.dart';
import '../widgets/common.dart';

class ProfileScreen extends StatefulWidget {
  const ProfileScreen({super.key});
  @override
  State<ProfileScreen> createState() => _ProfileScreenState();
}

class _ProfileScreenState extends State<ProfileScreen> {
  final _newUsername = TextEditingController();
  final _oldPw = TextEditingController();
  final _newPw = TextEditingController();
  bool _hasKey = false;
  bool _hasAvatar = false;
  String _visibility = 'public';
  int _bust = 0;
  bool _busy = false;
  String _error = '';
  String _message = '';

  String get _me => Auth.instance.user ?? '';

  @override
  void initState() {
    super.initState();
    Keystore.load(_me).then((id) => mounted ? setState(() => _hasKey = id != null) : null);
    getMe().then((me) {
      if (!mounted) return;
      setState(() {
        _hasAvatar = me['has_avatar'] == true;
        _visibility = (me['avatar_visibility'] as String?) ?? 'public';
      });
    }).catchError((_) {});
  }

  void _note(String m) => setState(() {
        _message = m;
        _error = '';
      });

  Future<void> _run(Future<void> Function() fn) async {
    setState(() {
      _busy = true;
      _error = '';
      _message = '';
    });
    try {
      await fn();
    } catch (e) {
      if (mounted) setState(() => _error = e.toString());
    } finally {
      if (mounted) setState(() => _busy = false);
    }
  }

  Future<void> _pickAvatar() async {
    final x = await ImagePicker().pickImage(source: ImageSource.gallery, imageQuality: 85, maxWidth: 1024);
    if (x == null) return;
    await _run(() async {
      await uploadAvatar(x.path);
      _hasAvatar = true;
      _bust++;
      _note('Profile photo updated');
    });
  }

  Future<void> _removeAvatar() => _run(() async {
        await deleteAvatar();
        _hasAvatar = false;
        _bust++;
        _note('Profile photo removed');
      });

  Future<void> _changeVisibility(String v) async {
    final prev = _visibility;
    setState(() => _visibility = v);
    try {
      await setAvatarVisibility(v);
    } catch (e) {
      setState(() {
        _visibility = prev;
        _error = e.toString();
      });
    }
  }

  Future<void> _changeUsername() async {
    final name = _newUsername.text.trim();
    if (name.isEmpty) return setState(() => _error = 'Please enter a new username');
    await _run(() async {
      await changeUsername(name);
      if (!mounted) return;
      await showDialog(
          context: context,
          builder: (c) => AlertDialog(title: const Text('Username updated'), content: const Text('Please sign in again.'),
              actions: [TextButton(onPressed: () => Navigator.pop(c), child: const Text('OK'))]));
      await Auth.instance.logout();
    });
  }

  Future<void> _changePassword() async {
    if (_oldPw.text.isEmpty || _newPw.text.isEmpty) return setState(() => _error = 'Please fill both password fields');
    if (_newPw.text.length < 8) return setState(() => _error = 'New password must be at least 8 characters');
    await _run(() async {
      Map<String, dynamic>? keys;
      try {
        // The server re-wraps with a fresh key pair on password change, as the
        // web and React Native clients do.
        if (_hasKey) keys = await wrapIdentity(generateIdentity(), _newPw.text);
      } catch (_) {}
      await changePassword(_oldPw.text, _newPw.text, keys);
      if (!mounted) return;
      await showDialog(
          context: context,
          builder: (c) => AlertDialog(title: const Text('Password updated'), content: const Text('Please sign in again.'),
              actions: [TextButton(onPressed: () => Navigator.pop(c), child: const Text('OK'))]));
      await Auth.instance.logout();
    });
  }

  Widget _section(String title, List<Widget> children) {
    final p = context.p;
    return Container(
      margin: const EdgeInsets.only(bottom: 12),
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(color: p.card, borderRadius: BorderRadius.circular(14), border: Border.all(color: p.border, width: 0.5)),
      child: Column(crossAxisAlignment: CrossAxisAlignment.stretch, children: [
        Text(title.toUpperCase(), style: TextStyle(color: p.subtext, fontSize: 12, fontWeight: FontWeight.w600, letterSpacing: 0.5)),
        const SizedBox(height: 12),
        ...children,
      ]),
    );
  }

  @override
  Widget build(BuildContext context) {
    final p = context.p;
    return Scaffold(
      appBar: AppBar(title: const Text('Profile', style: TextStyle(fontWeight: FontWeight.w700))),
      body: ListView(padding: const EdgeInsets.all(12), children: [
        if (_message.isNotEmpty) Padding(padding: const EdgeInsets.only(bottom: 8), child: Text(_message, style: TextStyle(color: p.success))),
        if (_error.isNotEmpty) Padding(padding: const EdgeInsets.only(bottom: 8), child: Text(_error, style: TextStyle(color: p.danger))),
        _section('Account', [
          Row(children: [
            UserAvatar(userId: _me, size: 64, bust: _bust),
            const SizedBox(width: 16),
            Expanded(
              child: Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
                Text(_me, style: TextStyle(fontSize: 18, fontWeight: FontWeight.w600, color: p.text)),
                Text(_hasKey ? '🔒 Encryption key on this device' : '⚠️ No encryption key on this device',
                    style: TextStyle(fontSize: 12, color: _hasKey ? p.secure : p.subtext)),
              ]),
            ),
          ]),
          const SizedBox(height: 12),
          Wrap(spacing: 8, children: [
            OutlinedButton(onPressed: _busy ? null : _pickAvatar, child: Text(_hasAvatar ? 'Change photo' : 'Add photo')),
            if (_hasAvatar) TextButton(onPressed: _busy ? null : _removeAvatar, child: Text('Remove', style: TextStyle(color: p.danger))),
          ]),
          const SizedBox(height: 8),
          Text('Who can see my photo', style: TextStyle(color: p.subtext, fontSize: 13)),
          const SizedBox(height: 6),
          SegmentedButton<String>(
            segments: const [
              ButtonSegment(value: 'public', label: Text('Everyone')),
              ButtonSegment(value: 'contacts', label: Text('Contacts')),
            ],
            selected: {_visibility},
            onSelectionChanged: (s) => _changeVisibility(s.first),
          ),
        ]),
        _section('Change username', [
          TextField(controller: _newUsername, autocorrect: false, decoration: const InputDecoration(labelText: 'New username')),
          const SizedBox(height: 10),
          PrimaryButton(label: 'Update username', busy: _busy, onPressed: _changeUsername),
        ]),
        _section('Change password', [
          TextField(controller: _oldPw, obscureText: true, decoration: const InputDecoration(labelText: 'Current password')),
          const SizedBox(height: 10),
          TextField(controller: _newPw, obscureText: true, decoration: const InputDecoration(labelText: 'New password')),
          const SizedBox(height: 10),
          PrimaryButton(label: 'Update password', busy: _busy, onPressed: _changePassword),
        ]),
        _section('Session', [
          Text('Server: $apiBase', style: TextStyle(color: p.subtext, fontSize: 12)),
          const SizedBox(height: 10),
          PrimaryButton(
            label: 'Sign out',
            color: p.danger,
            onPressed: () async {
              if (await confirm(context, 'Sign out?', 'Your encryption key is removed from this device.', ok: 'Sign out', destructive: true)) {
                await Auth.instance.logout();
              }
            },
          ),
        ]),
      ]),
    );
  }
}
